Added a bunch of TraitLocationAttributes

penev92 opened the pull request:
Follow-up of #19239
Also moved a few traits to their proper subfolders.
_EDIT: Also moved some traits from `^BaseWorld` to the `World` actor definition in TS rules._

~~There are still a couple of Jumpjet-related traits that I skipped since those were more questionable or I wasn't sure what to do with them. I can add those as well if we figure it out.~~ **EDIT: Handled below.**

penev92 committed:
Added a bunch of TraitLocationAttributes

Also moved a few traits to their proper subfolders.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/ExitsDebugOverlayManager.cs:
```diff
@@ -15,6 +15,7 @@
 namespace OpenRA.Mods.Common.Traits
 {
+	[TraitLocation(SystemActors.World | SystemActors.EditorWorld)]
 	public class ExitsDebugOverlayManagerInfo : TraitInfo
```
I don't think this makes sense to have in the editor?

penev92 replied:
Me neither but it's on ^BaseWorld :(
Shall I renove it from there? The editor showing a whole bunch of debug and effect visualizations is its own can of worms that we should tackle some day.

abcdefg30 replied:
You cannot enable this in the editor, so please move it to `World`.

abcdefg30 reviewed (commented):
Lgtm otherwise.

penev92 committed:
Moved some traits from ^BaseWorld to World in TS

penev92 commented:
Fixed.

Mailaender commented on OpenRA.Mods.Common/Traits/World/SubterraneanActorLayer.cs:
```diff
@@ -15,6 +15,7 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
+	[TraitLocation(SystemActors.World | SystemActors.EditorWorld)]
```
```suggestion
	[TraitLocation(SystemActors.World)]
```

imho as that is a runtime feature.

penev92 replied:
Good point, done.

Mailaender commented on OpenRA.Mods.Common/Traits/World/SubterraneanLocomotor.cs:
```diff
@@ -10,10 +10,12 @@
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Used by Mobile. Required for subterranean actors. Attach these to the world actor. You can have multiple variants by adding @suffixes.")]
+	[TraitLocation(SystemActors.World | SystemActors.EditorWorld)]
```
```suggestion
	[TraitLocation(SystemActors.World)]
```

This and `Locomotor` should actually not be required for mapping afaik.

pchote replied:
It may in the future for terrain passability information. We already have a usecase for this on aircraft (see recent discussion about aircraft being placed at cruise altitude vs landed).

penev92 replied:
Skipping to be consistent with other Locomotors.

penev92 commented:
- Moved `JumpjetActorLayer` and `SubterraneanActorLayer` from `^BaseWorld` to the `World` actor.
- Added a `TraitLocationAttribute` to `JumpjectLocomotor` to allow placement on `World` and `EditorWorld`.
- Added `TraitLocationAttribyte`s to `JumpjetActorLayer` and `SubterraneanActorLayer` to allow placement only on `World`.

phrohdoh reviewed (approved):
There is potential for future confusion in the interaction between this attribute and templated/inherited actors.

To test, in TS, I moved `JumpjetActorLayer` from `World` to `^BaseWorld` (from which `EditorWorld` inherits).
After doing so and running `make test` I see the following error (correctly failing `make test`):
```
OpenRA.Utility(1,1): Error: JumpjetActorLayer does not belong on editorworld. It is a system trait meant for World.
```
As a modder, I didn't put `JumpjetActorLayer` (directly) on `EditorWorld` so may I expect to see `^BaseWorld` mentioned instead.

Is this behavior (1) expected (are actor templates gone at this point?) and (2) best we can do currently (maybe forever)?

---

Approving due to `make test` failing as expected.
